### PR TITLE
Add Edge Authentication tests

### LINKED ISSUE #15: Add tests for Edge Authentication
Authorino can be used as an Authentication provider which will issue its own wristband token.

https://github.com/Kuadrant/authorino/blob/main/docs/user-guides/edge-authentication-architecture-festival-wristbands.md

- Test for the successful authorization with issued wristband token (200 OK)
- Test for the failed authorization with an incorrect token (401 Unauthorized)
- Test for the verification of information in the decoded wristband token

### PR DESCRIPTION
**Pull request by averevki**
- Move cfssl client to the root conftest
- Add Edge (Wristband) envoy template
- Add Edge Authentication tests
  - Test for the successful authorization with issued wristband token (200 OK)
  - Test for the failed authorization with an incorrect token (401 Unauthorized)
  - Test for the claims verification in the decoded wristband token

Closes #15

### CODE REVIEW

File: testsuite/tests/kuadrant/authorino/wristband/conftest.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            },
+        }
+    )
+    return authorization
+
+
+@pytest.fixture(scope="module")
+def wristband_token(client, auth):
+    """Test token acquirement from oidc endpoint"""
+    client.headers = {"Authorization": "Bearer " + auth.token.access_token}
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/conftest.py:65]: if you are using auth, you dont need to add headers manually, you can use `client.get("/auth", auth=auth)`

File: testsuite/tests/kuadrant/authorino/wristband/conftest.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    """Client for second AuthConfig"""
+    client = envoy.client()
+    client.base_url = f"http://{hostname2}"
+    yield client
+    client.close()
+
+
+# pylint: disable=unused-argument
+@pytest.fixture(scope="module", autouse=True)
+def commit(request, commit, certificates, authorization2):
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/conftest.py:107]: I am not sure certificates need to be there, they should be requested automatically either way

File: testsuite/tests/kuadrant/authorino/wristband/test_wristband.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+def test_wristband_token_claims(wristband_token, wristband_endpoint, certificates):
+    """Verify acquired jwt token claims"""
+    # extract public key from the certificate
+    cert_encoded = certificates["signing_ca"].certificate.encode("utf-8")
+    cert_object = load_pem_x509_certificate(cert_encoded)
+    public_key = cert_object.public_key()
+    # decode jwt to examine claims
+    wristband_decoded = jwt.decode(wristband_token, key=public_key, algorithms=["RS256"])
+    assert wristband_decoded["exp"] - wristband_decoded["iat"] == 300
+    assert wristband_decoded["iss"] == wristband_endpoint
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/test_wristband.py:16]: Could you also test that the token contains less information than the original one from RHSSO?

File: testsuite/tests/kuadrant/authorino/wristband/conftest.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+@pytest.fixture(scope="module")
+def envoy_template():
+    """Envoy with additional route match"""
+    return "envoy_wristband.yaml"
+
+
+@pytest.fixture(scope="module")
+def wristband_endpoint(openshift, authorino, authorization_name):
+    """Authorino oidc wristband endpoint"""
+    return f"http://{authorino.oidc_url}:8083/{openshift.project}/{authorization_name}/wristband"
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/conftest.py:37]: Do you need to specify the openshift project?

File: testsuite/tests/kuadrant/authorino/wristband/conftest.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
+"""Conftest for Edge Authentication tests"""
+import pytest
+
+from testsuite.openshift.objects.auth_config import AuthConfig
+from testsuite.certificates import CertInfo
+from testsuite.objects import Authorization
+from testsuite.utils import cert_builder
+
+
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/conftest.py:9]: There is a missing `run_on_kuadrant` fixture (should return False), as you need custom envoy

File: testsuite/tests/conftest.py [hunk 6]
```diff
@@ -235,12 +235,20 @@ def backend(request, openshift, blame, label):
 
 
 @pytest.fixture(scope="module")
-def envoy(request, kuadrant, authorino, openshift, blame, backend, module_label, testconfig) -> Proxy:
+def envoy_template():
```
Comment (pehala) [inline testsuite/tests/conftest.py:238]: I would not do this in root conftest, we are trying to avoid using envoy modification as much as possible, I would override envoy fixture in wristband conftest instead.

File: testsuite/tests/kuadrant/authorino/conftest.py [hunk 7]
```diff
@@ -64,6 +65,15 @@ def auth(oidc_provider):
     return HttpxOidcClientAuth(oidc_provider.get_token, "authorization")
 
 
+@pytest.fixture(scope="session")
+def cfssl(testconfig):
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/conftest.py:69]: You can move it up to the root conftest in this particular case

File: testsuite/tests/kuadrant/authorino/wristband/conftest.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    assert response.headers.get("wristband-token") is not None
+    return response.headers["wristband-token"]
+
+
+# pylint: disable=unused-argument
+@pytest.fixture(scope="module")
+def authorization2_name(blame, oidc_provider):
+    """Second AuthConfig unique name"""
+    return blame("authz2")
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/conftest.py:77]: Do you need the name in multiple fixtures? If not you can generate it only where it is needed.

File: testsuite/tests/kuadrant/authorino/wristband/conftest.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+
+@pytest.fixture(scope="module")
+def hostname2(envoy, blame):
+    """Second route for the envoy"""
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/conftest.py:82]: Which route is it? It would be great if it was easy to see if this is the wristband route or the other one.

File: testsuite/openshift/client.py [hunk 10]
```diff
@@ -149,6 +149,22 @@ def create_tls_secret(self, name: str, certificate: Certificate, labels: Optiona
         with self.context:
             return oc.create(model, ["--save-config=true"])
 
+    def create_signing_secret(self, name: str, key: str, labels: Optional[Dict[str, str]] = None):
```
Comment (pehala) [inline testsuite/openshift/client.py:152]: Would it make sense to update the existing create_tls_secret method? e.g. ability to rename the keys?

File: testsuite/tests/kuadrant/authorino/wristband/conftest.py (hunk 9 above)
Comment (pehala) [reply testsuite/tests/kuadrant/authorino/wristband/conftest.py:89]: What does "core" mean? I would appreciate a better docstring

File: testsuite/tests/kuadrant/authorino/wristband/conftest.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    authorization2 = AuthConfig.create_instance(
+        openshift, blame("auth-core"), None, hostnames=[hostname_core], labels={"testRun": module_label}
+    )
+    authorization2.identity.oidc("edge-authenticated", wristband_endpoint)
+    return authorization2
+
+
+@pytest.fixture(scope="module")
+def client2(hostname_core, envoy):
+    """Client for second AuthConfig"""
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/conftest.py:105]: I would again opt for better naming that reflects more its purpose, in both name and docstring

File: testsuite/tests/kuadrant/authorino/wristband/test_wristband.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+from jose import jwt
+
+
+def test_wristband_token_claims(oidc_provider, auth, wristband_token, wristband_endpoint, certificates):
+    """Verify acquired jwt token claims"""
+    wristband_decoded = jwt.decode(wristband_token, certificates["signing_ca"].certificate)
+    assert wristband_decoded["exp"] - wristband_decoded["iat"] == 300
+    assert wristband_decoded["iss"] == wristband_endpoint
+    # check differences in claims between rhsso token and acquired wristband token
+    access_token_decoded = jwt.decode(auth.token.access_token, oidc_provider.get_public_key(), audience="account")
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/test_wristband.py:11]: There is a `decode_token` method in the python-keycloak library

Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/test_wristband.py:11]: By using `get_public_key()` without specifically requesting RHSSO, this test will fail if we switch to auth0, as the method is not there

Comment (averevki) [reply testsuite/tests/kuadrant/authorino/wristband/test_wristband.py:11]: This method uses a hard-coded audience claim check as client id, which does not match the actual audience value (it is a known issue https://stackoverflow.com/questions/53550321/keycloak-gatekeeper-aud-claim-and-client-id-do-not-match). There is an option to not verify audience and use this build-in method, but I would prefer to decode it by using the jwt library (as it is now).

File: testsuite/openshift/client.py (hunk 10 above)
Comment (averevki) [reply testsuite/openshift/client.py:152]: I would prefer to leave these secrets as they are for now. Because, in my opinion, they have a lot of differences, such as different types, key names, count of secret pairs.

Comment (pehala) [reply testsuite/openshift/client.py:152]: Would it work with both key and cert in the secret? If yes, then I think adding the option to rename the key and change the type would be enough and it wouldn't be that different

**Comment by averevki** [comment]
rebase

File: testsuite/openshift/client.py [hunk 13]
```diff
@@ -132,14 +132,29 @@ def is_ready(self, selector: Selector):
...
 
-    def create_tls_secret(self, name: str, certificate: Certificate, labels: Optional[Dict[str, str]] = None):
-        """Creates a TLS secret"""
+    def create_key_secret(
+        self,
+        name: str,
+        certificate: Certificate,
+        type_: Literal["kubernetes.io/tls", "Opaque"],
+        labels: Optional[Dict[str, str]] = None,
+    ):
```
Comment (pehala) [inline testsuite/openshift/client.py:141]: I would add two parameters: key_name, cert_name which we would use as keys for the secret and wouldn't make it bound to the specific type

File: testsuite/tests/kuadrant/authorino/wristband/test_wristband.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    for claim in ["preferred_username", "email", "realm_access", "resource_access"]:
+        assert claim in access_token_decoded
+        assert claim not in wristband_decoded
+
+
+def test_wristband_success(client_authenticated, wristband_token):
+    """Test successful Edge Authentication"""
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/wristband/test_wristband.py:18]: Please explain what the test tests and why it should succeed

File: testsuite/tests/kuadrant/authorino/operator/tls/conftest.py [hunk 15]
```diff
@@ -62,22 +62,15 @@ def create_secret(blame, request, openshift):
 
     def _create_secret(certificate: Certificate, name: str, labels: Optional[Dict[str, str]] = None):
         secret_name = blame(name)
-        secret = openshift.create_tls_secret(secret_name, certificate, labels=labels)
+        secret = openshift.create_tls_secret(
+            secret_name, certificate, "tls.crt", "tls.key", "kubernetes.io/tls", labels=labels
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/tls/conftest.py:66]: You can have tls.crt, tls.key and "kubernetes.io/tls" as defaults

**Review by jsmolar — APPROVED** [review]
